src/CommandLine/Core/InstanceBuilder.cs

```diff
@@ -28,13 +28,15 @@ static class InstanceBuilder
             var typeInfo = factory.MapValueOrDefault(f => f().GetType(), typeof(T));
 
             var specProps = typeInfo.GetSpecifications(pi => SpecificationProperty.Create(
-                    Specification.FromProperty(pi), pi, Maybe.Nothing<object>()));
+                    Specification.FromProperty(pi), pi, Maybe.Nothing<object>()))
+                .Memorize();

```

Comment:
I used .Memorize() throughout this PR since that seems to be the dominant existing convention.

However, it feels like an anti-pattern to me to pass IEnumerable everywhere rather than, say, IReadOnlyCollection. It makes it difficult to detect multi-enumeration issues visually or via static analysis such as Resharper. IEnumerable definitely has its place, it just feels overused.

What is the rationale behind the current approach?

Reply:
> What is the rationale behind the current approach?

`IReadOnlyCollection` was introduced in .NET 4.5. This library still targets 4.0 on the fat framework.

That, plus the developer was really inspired by functional languages and the "lazy evaluation" that's built in to many of them. Unfortunately, there are (perf) penalties for using C# as if it were a functional language.